Trigger async Eden GC in OTS where appropriate

Andrew-Fryer opened the pull request:
#### 8403d0ef0aa9be07ec326935567c5c6088732e5b
<pre>
Trigger async Eden GC in OTS where appropriate
<a href="https://bugs.webkit.org/show_bug.cgi?id=296507">https://bugs.webkit.org/show_bug.cgi?id=296507</a>
<a href="https://rdar.apple.com/156748517">rdar://156748517</a>

Reviewed by Yusuke Suzuki.

OTS already estimates GC length to determine if it should trigger GC.
In the case that there isn&apos;t quite enough time, it is best to trigger
async GC if we&apos;d otherwise trigger GC in slow allocation.
This way, we&apos;ll complete most of the GC during idle time rather than
having to complete an entire GC while we&apos;re busy.

* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::performOpportunisticallyScheduledTasks):

Canonical link: <a href="https://commits.webkit.org/298045@main">https://commits.webkit.org/298045@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be2f30260adeca41d68b2633a02777a530b8f9b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113241 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23454 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119449 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64484 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41539 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86187 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41515 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116188 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101865 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66509 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26130 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19993 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63203 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105758 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96240 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20068 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122666 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111857 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40319 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30077 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95035 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40704 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98080 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94777 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24290 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39925 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17740 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36455 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40205 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45704 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136087 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39846 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36517 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43179 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41583 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/be2f30260adeca41d68b2633a02777a530b8f9b4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48547

rreno commented on Source/JavaScriptCore/runtime/VM.cpp:
```diff
@@ -1760,6 +1760,11 @@ void VM::performOpportunisticallyScheduledTasks(MonotonicTime deadline, OptionSe
                 heap.m_shouldDoOpportunisticFullCollection = false;
                 return;
+            } else if (estimatedGCDuration < 2 * remainingTime) {
+                if (heap.totalBytesAllocatedThisCycle() * 2 > heap.m_minBytesPerCycle) {
```
Is this the same value used by slow allocations to decide if they need to trigger a GC? If so, this should probably be promoted to a well-named function on Heap (or wherever it needs to go) so that this calculation and the one in slow allocations stay in sync. I can imagine a tweak in the future to slow allocation heuristics blowing this thing up unexpectedly (or causing it to do something other than intended, at least).

Andrew-Fryer replied:
I think it's fine to access the member variable directly because we're already doing that in the surrounding code (allowed by the fact that Heap and VM are friend classes).
I actually originally wrote the patch with a getter and then figured it was actually cleaner to remove it.
If anyone has strong opinions here I can add getters (or similar methods) for the other member variables accessed here too.

Andrew-Fryer replied:
To answer your question: No.
The slow allocation check uses m_maxHeapSize which is significantly different (allows for eden to become larger in most cases).
The benefit of m_minBytesPerCycle is that it is constant (after VM initialization), which makes it simpler in a way.
(I plan to experiment with changing this heuristic in the future.)

Andrew-Fryer committed:
Trigger async Eden GC in OTS where appropriate
https://bugs.webkit.org/show_bug.cgi?id=296507
rdar://156748517

Reviewed by Yusuke Suzuki.

OTS already estimates GC length to determine if it should trigger GC.
In the case that there isn't quite enough time, it is best to trigger
async GC if we'd otherwise trigger GC in slow allocation.
This way, we'll complete most of the GC during idle time rather than
having to complete an entire GC while we're busy.

* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::performOpportunisticallyScheduledTasks):

Canonical link: https://commits.webkit.org/298045@main

webkit-commit-queue commented:
Committed 298045@main (8403d0ef0aa9): <https://commits.webkit.org/298045@main>

Reviewed commits have been landed. Closing PR #48547 and removing active labels.
